assayist/processor/container_go_analyzer.py

```diff
@@ -0,0 +1,147 @@
+        if returncode:
+            raise RuntimeError(f'The command "{" ".join(cmd)}" failed with: {stderr}')
+
+        # There may be many top-level repositories, but we are only
+        # allowed a single 'upstream' relationship type. Only report
+        # the first top-level repository as "the upstream" one.
+        seen_toplevel = False
+
+        # Parse the output from backvendor.
+        for line in stdout.splitlines():
+            fields = line.split('\t')
+            if len(fields) != 4:
+                log.error(f'invalid backvendor output: {line}')
+                continue
+
+            mod, ver, repo, rev = fields
+
+            # The module field begins with an asterisk for a top-level module
+            srctype = self.VENDORED
+            if mod.startswith('*'):
+                mod = mod[1:]
+                if not seen_toplevel:
+                    srctype = self.UPSTREAM
+                    seen_toplevel = True
+
+            self._process_go_module(source_location, srctype,
+                                    mod, ver, repo, rev)
+
+    def _process_go_module(self, source_location, srctype,
+                           mod, ver, repo, rev):
+        """Update the database with the information from backvendor.
+
+        :param SourceLocation source_location: local source location DB node
+        :param str srctype: either self.UPSTREAM or self.VENDORED
+        :param str mod: module name
+        :param str ver: version
+        :param str repo: repository URL
+        :param str rev: vcs revision
+        """
+        upstream = self.create_or_update_source_location("#".join(repo, rev),
+                                                         ver)
+        relationship = getattr(source_location, srctype)
+        relationship.connect(upstream)
+
+        namespace, name = mod.split('/', 1)  # There must be at least one '/'
+        component = source.Component.get_or_create({
+            'canonical_type': 'golang',
+            'canonical_namespace': namespace,
+            'canonical_name': name})[0]
+
+        # Connect the component to the upstream source location
+        component.source_locations.connect(upstream)
+
+        if srctype == self.UPSTREAM:
+            # Connect the component to the local source location it is
+            # the upstream for.
+            component.source_locations.connect(source_location)
+
+    def _claim_go_executables(self):
+        not_container_msg = 'Skipping archive {0} since it\'s not a container image'
+        for archive in self.read_metadata_file(self.ARCHIVE_FILE):
+            if not self.is_container_archive(archive):
+                log.debug(not_container_msg.format(archive['id']))
+                continue
+
+            layer_dir = os.path.join(self.input_dir, 'unpacked_archives',
+                                     'container_layer',
+                                     archive['filename'])
+
+            cmd = [self.GOVERSION, '.']
```

Comment:
Running this will just give us the go binaries, but not necessarily the go binaries produced from the embedded source locations. Is this a risk we can take?

Reply:
I could add a heuristic to this: only match files whose names are the default binary names Go would produce from each package (essentially the package's basename). WDYT?

Reply:
@twaugh if you think that's reliable, then that sounds good.

Reply:
I think it's reliable enough Go 1.11 comes along to allow goversion to provide richer information from the executables.

Reply:
This isn't going to work. For example, kubevirt gets installed as '/usr/bin/virt-api'.

Until we move to Go modules built with Go 1.12+ we are not going to get sufficient information to be able to identify these conclusively.

I'll remove the basename check.

Reply:
Removed.